### PR TITLE
Create infrastructure for experiments.

### PR DESCRIPTION
**Pull request by Mitten-O**
Often, especially when tweaking gigi, it is annoying to have
to link the entire freeorion binary and then maybe even load a
game to gain access to the dialog whose underlying controls you were
changing.

With this pull request, you can make small test applications for specific
things in a special folder. Included is TestTextControl.cpp, which shows a small
program that can be used to see how text is rendering.
The experiments won't be built unless you explicitly request it.

### CODE REVIEW

File: experiments/GG/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,55 @@
+cmake_minimum_required(VERSION 2.6)
+
+message("-- Configuring GG Tests")
+
+find_package(OpenAL REQUIRED)
+find_package(Ogg REQUIRED)
+find_package(Vorbis REQUIRED)
+find_package(Bullet REQUIRED)
```
Comment (geoffthemedio) [inline experiments/GG/CMakeLists.txt]: why bullet?

File: experiments/GG/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+set(gg_test_LINK_LIBS
+    GiGi
+    GiGiSDL
+    ${OPENGL_gl_LIBRARY}
+    ${OPENGL_glu_LIBRARY}
+    ${OPENAL_LIBRARY}
+    ${Boost_LIBRARIES}
+    ${OGG_LIBRARY}
+    ${VORBIS_LIBRARIES}
+    ${BULLET_LIBRARIES}
```
Comment (geoffthemedio) [inline experiments/GG/CMakeLists.txt]: bullet?

**Comment by Mitten-O** [comment]
Oh sorry about the bullet, I've actually had this for myself since before the SDL migration and related purges. I missed cleaning those up. Did that now.

Warning: rebased pull request branch.

File: experiments/GG/TestApp.cpp [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+#include <GG/SDL/SDLGUI.h>
+
+#include <boost/lexical_cast.hpp>
+#include <boost/format.hpp>
+#include <boost/filesystem/operations.hpp>
+#include <boost/filesystem/fstream.hpp>
+
+class MinimalGGApp : public GG::SDLGUI {
+    public:
+        MinimalGGApp (int width, int height, bool calculate_FPS, const std::string& name, int x, int y, bool fullscreen, bool fake_mode_change);
```
Comment (geoffthemedio) [inline experiments/GG/TestApp.cpp:19]: please linebreak the long list of parameters, and remove the space in "App ("

File: experiments/GG/TestApp.cpp [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+{
+    boost::shared_ptr<GG::Texture> cursor_texture = GG::GetTextureManager().GetTexture(
+        (boost::filesystem::path("../../default") / "data" / "art" / "cursors" / "default_cursor.png").string(), false);
+    SetCursor(boost::shared_ptr<GG::TextureCursor>(new GG::TextureCursor(cursor_texture, GG::Pt(GG::X(6), GG::Y(3)))));
+    RenderCursor(true);
+
+    GLInit();
+}
+
+MinimalGGApp::~MinimalGGApp() {
```
Comment (geoffthemedio) [inline experiments/GG/TestApp.cpp:44]: change to

```
MinimalGGApp::~MinimalGGApp()
{}
```

File: experiments/GG/TestApp.cpp [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+// GG is not rendering "in 2D" (orthographic, with coordinates mapped 1-1 onto
+// the pixels of the display).  To move into "2D" mode, GG calls Enter2DMode()
+// at the beginning of each GG rendering iteration.  This method is
+// responsible for preserving any state it alters, then altering the necessary
+// state to match what GG requires.  GG requires GL_TEXTURE_2D to be on,
+// GL_LIGHTING off, and requires a viewport and orthographic projection as
+// shown below.  The other code you see here is usually convenient to working
+// with GG, but not required.
+void MinimalGGApp::Enter2DMode()
+{
```
Comment (geoffthemedio) [inline experiments/GG/TestApp.cpp:58]: put { at end of a single-line function definition starting...

File: experiments/GG/TestApp.cpp [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+    } catch (const std::exception& e) {
+        std::cerr << "main() caught exception(std::exception): " << e.what() << std::endl;
+    } catch (...) {
+        std::cerr << "main() caught unknown exception." << std::endl;
+    }
+}
+
+TestAppImpl::~TestAppImpl() {
+    delete m_app;
```
Comment (geoffthemedio) [inline experiments/GG/TestApp.cpp:215]: could do

```
TestAppImpl::~TestAppImpl()
{ delete m_app; }
```

 for single-line functions

File: experiments/GG/TestApp.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    /// An option database will be initialized from them.
+    /// You need to register your options.
+    TestApp(int argc, char** argv, unsigned width = 400, unsigned height = 300);
+
+    /// The given window will be made visible.
+    /// Then the event pump is started.
+    /// This method only returns once the application quits
+    void Run(GG::Wnd* wnd);
+
+    private:
```
Comment (geoffthemedio) [inline experiments/GG/TestApp.h:22]: remove extra indentation here

File: experiments/GG/TestDialog.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+void TestDialog::Render() {
+    GG::Wnd::Render();
+    GG::FlatRectangle ( UpperLeft(), LowerRight(),
+                        GG::CLR_GRAY, GG::CLR_SHADOW, 1u );
+}
+
+GG::Pt TestDialog::ClientLowerRight() const {
+    return LowerRight() - GG::Pt(GG::X(10), GG::Y(10) );
```
Comment (geoffthemedio) [inline experiments/GG/TestDialog.cpp:23]: single line functions can have their { /_code_/ } on a single line...

File: experiments/GG/TestTextControl.cpp [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+#include <GG/StyleFactory.h>
+
+#include "TestApp.h"
+
+int main(int argc, char* argv[])
+{
+
+    TestApp app(argc, argv);
+
+    const std::string message = "Are <i>we</i> Готово yet?"; // That Russian word means "Done", ha.
```
Comment (geoffthemedio) [inline experiments/GG/TestTextControl.cpp:14]: I think you might need some extra stuff before/after the "" to make this work properly as a UTF-8 string constant.

File: experiments/GG/TestTextControl.cpp [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    const std::set<GG::UnicodeCharset> charsets_ = GG::UnicodeCharsetsToRender(message);
+    const std::vector<GG::UnicodeCharset> charsets(charsets_.begin(), charsets_.end());
+
+    const boost::shared_ptr<GG::Font> font =
+    GG::GUI::GetGUI()->GetStyleFactory()->DefaultFont(12, &charsets[0], &charsets[0] + charsets.size());
+
+    GG::Wnd* quit_dlg =
+    new GG::ThreeButtonDlg(GG::X(10), GG::Y(10), message, font, GG::CLR_SHADOW,
```
Comment (geoffthemedio) [inline experiments/GG/TestTextControl.cpp:22]: indent continued line by one step (4 spaces)

**Comment by Mitten-O** [comment]
Thanks for bothering to give such detailed input. I think I've addressed all of your concerns now. Had to use a c++11 feature to ensure the message gets encoded as UTF-8, but since this is development time only stuff, that should be ok?

Warning: rebased.

**Comment by Vezzra** [comment]
> Had to use a c++11 feature to ensure the message gets encoded as UTF-8, but since this is development time only stuff, that should be ok?

Well, can't be used on OSX until I figure out how to get FO to compile with C++11 enabled again. Won't happen before 0.4.5 is out, though.

**Comment by Mitten-O** [comment]
Well, the test program and the helpers for it are just a few dozen lines. They are happy to be linked to GiGi produced with entirely traditional options and have no idea about FO in general, so whether or not the rest of the environment is using c++11 should be of little consequence. Should. Never know with c++, though.

It's little enough trouble to clean up the _u8_ and the _override_s if someone feels like keeping the code base clean, I just figured that in code only used for faster toying by devs doesn't need to care about the general public, and should take advantage of all short cuts available.

I do welcome input on the general usefulness of this pull request, also. As I mentioned, I've had this for myself for a while and figured someone else may find it useful, but it's easy enough to splice out the existence of this from my public commits if nobody else desires faster GiGi experimentation feedback loops.

**Comment by Dilvish-fo** [comment]
> The experiments won't be built unless you explicitly request it.

And furthermore, currently handling for that is only provided via cmake, so it seems having this C++11 dependence shouldn't cause any problems for a regular FO build on MacOSX; hopefully we'll get the C++11 issues sorted out there in the not too distant future.

**Comment by Vezzra** [comment]
> hopefully we'll get the C++11 issues sorted out there in the not too distant future.

Looks like I got myself an assignment... :wink:

File: experiments/GG/CMakeLists.txt (hunk 2 above)
Comment (geoffthemedio) [inline experiments/GG/CMakeLists.txt]: bullet references still there...

Comment (Mitten-O) [inline experiments/GG/CMakeLists.txt]: I am certain I removed them at some point .. I must have messed up some rebase.

**Comment by Vezzra** [comment]
@adrianbroher, shouldn't the milestone be adjusted accordingly?

**Pull request closed without merging**